pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/StUnionAggregationFunction.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.pinot.core.query.request.context.ExpressionContext;
+import org.apache.pinot.spi.utils.ByteArray;
+import org.locationtech.jts.geom.Geometry;
+
+
+public class StUnionAggregationFunction extends BaseSingleInputAggregationFunction<Geometry, ByteArray> {
+
+  /**
+   * Constructor for the class.
+   *
+   * @param expression Expression to aggregate on.
+   */
+  public StUnionAggregationFunction(ExpressionContext expression) {
+    super(expression);
+  }
+
+  @Override
+  public AggregationFunctionType getType() {
+    return AggregationFunctionType.ST_UNION;
+  }
+
+  @Override
+  public void accept(AggregationFunctionVisitorBase visitor) {
+    visitor.visit(this);
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return new ObjectAggregationResultHolder();
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return new ObjectGroupByResultHolder(initialCapacity, maxCapacity);
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    byte[][] bytesArray = blockValSetMap.get(_expression).getBytesValuesSV();
+    Geometry geometry = aggregationResultHolder.getResult();
+    for (int i = 0; i < length; i++) {
+      Geometry value = GeometrySerializer.deserialize(bytesArray[i]);
+      geometry = geometry == null ? value : geometry.union(value);
+    }
+    aggregationResultHolder.setValue(geometry);
+  }
+
+  @Override
+  public void aggregateGroupBySV(int length, int[] groupKeyArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    byte[][] bytesArray = blockValSetMap.get(_expression).getBytesValuesSV();
+    for (int i = 0; i < length; i++) {
+      int groupKey = groupKeyArray[i];
+      Geometry value = GeometrySerializer.deserialize(bytesArray[i]);
+      groupByResultHolder.setValueForKey(groupKey, groupByResultHolder.getResult(groupKey) == null ? value
+          : ((Geometry) groupByResultHolder.getResult(groupKey)).union(value));
+    }
+  }
+
+  @Override
+  public void aggregateGroupByMV(int length, int[][] groupKeysArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    byte[][] bytesArray = blockValSetMap.get(_expression).getBytesValuesSV();
+    for (int i = 0; i < length; i++) {
+      Geometry value = GeometrySerializer.deserialize(bytesArray[i]);
+      for (int groupKey : groupKeysArray[i]) {
+        groupByResultHolder.setValueForKey(groupKey, groupByResultHolder.getResult(groupKey) == null ? value
+            : ((Geometry) groupByResultHolder.getResult(groupKey)).union(value));
+      }
+    }
+  }
+
+  @Override
+  public Geometry extractAggregationResult(AggregationResultHolder aggregationResultHolder) {
+    return aggregationResultHolder.getResult();
```

Comment:
Need to handle `null` here

Reply:
Can you explain the handling here, throw exception or something? My understanding is that the caller shall do the null handling?

Reply:
If there is no record selected, then `aggregationResultHolder.getResult();` will return null because there is no value set yet. In that case, you should return an empty geometry as the result of the current segment

Reply:
My question is why not use null as the placeholder instead of an empty abstract geometry? For example, in line #86 I checked if the group by result is null.

Perhaps I miss something in the general null handling strategy in agg functions.

Reply:
It won't work because `DataTable` does not support `null` as input (for sending back response from server to broker)

Reply:
ah, I see. So all aggregation result and group by result cannot use null as holders.

Reply:
The holder might contain `null`, but before returning the segment result, we need to convert `null` to the default value

Reply:
yup. Makes sense to me now